src/propagators.jl

```diff
@@ -0,0 +1,208 @@
+using QuBase
+using Docile
+
+import Base: start, next, done
+
+abstract QuPropagatorMethod
+
+immutable QuPropagator{QPM<:QuPropagatorMethod}
+    hamiltonian
+    init_state::QuVector
+    tlist
+    method::QPM
+    QuPropagator(hamiltonian, init_state, tlist, method) = new(hamiltonian, init_state, tlist, method)
+end
+
+QuPropagator{QPM<:QuPropagatorMethod}(hamiltonian, init_state, tlist, method::QPM) = QuPropagator{QPM}(hamiltonian, init_state, tlist, method)
+
+immutable QuPropagatorState
+    psi::QuVector
+    t
+    t_state
+end
+
+@doc """
+Input Parameters : QuPropagator
+
+Returns the starting, i.e., the initial state of the system
+""" ->
+function Base.start(prob::QuPropagator)
+    t_state = start(prob.tlist)
+    t,t_state = next(prob.tlist,t_state)
+    return QuPropagatorState(prob.init_state,t,t_state)
+end
+
+@doc """
+Input Parameters : QuPropagator and QuPropagator State
+
+Returns the next state by dispatching to particular
+`propagate` method depending on the `numerical` method type.
+""" ->
+function Base.next{QPM<:QuPropagatorMethod}(prob::QuPropagator{QPM}, qustate::QuPropagatorState)
+    op = prob.hamiltonian
+    current_qustate = qustate.psi
+    current_t = qustate.t
+    if done(prob, qustate) == false
```

Comment:
I think you don't need that test at all. If `done` would be `true` the new time `t` shouldn't change (at least this is my understanding of the iterator protocol) and thus `t-current_t==0`. Moreover, in the current form the function is not type-stable, since you return `(state, time), propagator state` and `propagator state` depending on the **value** `qustate`. So I would simply remove the `if ... else ... end` construct. 

Also not that you return `(state, time)` instead of `(time,state)` (I don't remember if we specified this at some point, but the latter is more consistent with ODE.jl).